niceml/data/datasets/dfdataset.py

```diff
@@ -121,24 +256,68 @@ def __len__(self):
     def on_epoch_end(self):
         """shuffles on epoch end"""
         if self.shuffle:
-            random.shuffle(self.index_list)
+            self.index_list = self.data_shuffler.shuffle(
+                data_infos=self.get_all_data_info()
+            )
 
     def iter_with_info(self):
+        """
+        The iter_with_info function is a generator that yields the next batch of data,
+        along with some additional information about the batch.
+        The additional information is useful for various diagnostic purposes.
+        The function returns an object of type DataIterator, which has two fields:
+            * 'batch' contains the next batch of data.
+            * 'info' contains additional information about that batch.
+
+        Returns:
+            A dataiterator object
+        """
         return DataIterator(self)
 
     def get_datainfo(self, batch_index):
         """creates datainfo for given index"""
         start_idx = batch_index * self.batch_size
         end_idx = min(len(self.index_list), (batch_index + 1) * self.batch_size)
         data_info_list: List[RegDataInfo] = []
-        for cur_idx in range(start_idx, end_idx):
-            real_index = self.index_list[cur_idx]
-            data_info_dict = {}
-            for cur_data in self.inputs + self.targets:
-                data_info_dict[cur_data["key"]] = self.dataframe.loc[
-                    real_index, cur_data["key"]
-                ]
-            cur_id = self.dataframe.loc[real_index, self.id_key]
-            data_info_dict = json.loads(json.dumps(data_info_dict))
-            data_info_list.append(RegDataInfo(cur_id, data_info_dict))
+        input_keys = [input_dict["key"] for input_dict in self.inputs]
+        target_keys = [target_dict["key"] for target_dict in self.targets]
+        data_subset = self.data[
+            [self.id_key] + input_keys + target_keys + self.extra_key_list
+        ]
+        real_index_list = [self.index_list[idx] for idx in range(start_idx, end_idx)]
+        data_info_dicts: List[dict] = data_subset.iloc[real_index_list].to_dict(
+            "records"
+        )
+
+        for data_info_dict in data_info_dicts:
+            key = data_info_dict[self.id_key]
+            data_info_dict.pop(self.id_key)
+            data_info_list.append(RegDataInfo(key, data_info_dict))
+        return data_info_list
+
+    def get_all_data_info(self) -> List[RegDataInfo]:
+        """
+        The get_all_data_info function returns a list of RegDataInfo objects.
+        Each RegDataInfo object contains the following information:
```

Comment:
I think the detailed information about what the RegDataInfo contains should be given at the RegDataInfo class itself. Please remove this from the functions docstring.